### PR TITLE
[1/2] Use any task as an embedder for any layer

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

This PR adds support for calling `as_embedder` on any task to convert it to an embedder. The embeddings retain gradient information so they could be used for training the model weights.

This approach uses an exception to break out of the forward in intermediate layers to avoid excess computation.

Still to come in follow-up: docs

## Before submitting
- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? - Coming in 2/2
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [x] Did you verify **new and existing tests pass** locally with your changes?
- [x] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_model.py [hunk 1]
```diff
@@ -290,6 +291,32 @@ def test_model_download(tmpdir, cls, filename):
...
+
+    def predict_step(self, batch, batch_idx: int, dataloader_idx: int = 0):
+        return self.model(batch)
+
+
+def test_as_embedder():
+    embedder = DummyTask().as_embedder("model.1")
+
+    assert isinstance(embedder, Embedder)
+    assert embedder.predict_step(torch.rand(10, 10), 0, 0).size(1) == 30
```
Comment (krshrimali) [inline tests/core/test_model.py:312]: Minor suggestion: probably better if we don't hard-code, and use `out_features` from the respective layer:

```suggestion
def test_as_embedder():
    layer_number = "1"
    embedder = DummyTask().as_embedder("model." + layer_number)

    assert isinstance(embedder, Embedder)
    assert (
        embedder.predict_step(torch.rand(10, 10), 0, 0).size(1)
        == embedder.model.model._modules[layer_number].out_features
    )
```

File: tests/core/test_model.py [hunk 2]
```diff
@@ -290,6 +291,33 @@ def test_model_download(tmpdir, cls, filename):
         assert isinstance(task, cls)
 
 
+class DummyTask(Task):
+    def __init__(self):
+        super().__init__()
+
+        self.model = nn.Sequential(
```
Comment (justusschock) [inline tests/core/test_model.py:298]: very nit: for testing: Can we use another attribute name here (not `self.model`)? This would ease debugging since it is also called `self.model` inside the embedder class.

File: tests/core/test_model.py [hunk 3]
```diff
@@ -290,6 +291,33 @@ def test_model_download(tmpdir, cls, filename):
...
+            nn.Linear(30, 40),
+        )
+
+    def predict_step(self, batch, batch_idx: int, dataloader_idx: int = 0):
+        return self.model(batch)
+
+
+def test_as_embedder():
+    layer_number = "1"
+    embedder = DummyTask().as_embedder(f"model.{layer_number}")
```
Comment (justusschock) [inline tests/core/test_model.py:310]: just a question: Should we also allow to only call `.as_embedder(layer_number)`? I.e. splitting after the last separator '.' and check those as well? I think almost all builtin tasks only have a single model, right?

**Review by justusschock — APPROVED** [review]
LGTM

File: flash/core/utilities/embedder.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    def _register_hook(self):
+        if self._module is not None:
+            self._handle = self._module.register_forward_hook(self._hook)
+
+    def _remove_hook(self):
+        if self._handle is not None:
+            self._handle.remove()
+        self._handle = None
+
+    def training_step(self, batch: Any, batch_idx: int) -> Any:
```
Comment (otaj) [inline flash/core/utilities/embedder.py:65]: Are you sure we don't to be training embedders?

**Review by otaj — APPROVED** [review]
LGTM!

File: flash/core/utilities/embedder.py (hunk 4 above)
Comment (ethanwharris) [reply flash/core/utilities/embedder.py:65]: Not trainable for now :p Of course, you could always override the training step here to implement some custom logic for training with embeddings, but it shouldn't run without having any loss or optimizer etc. so needs some kind of error here.

File: tests/core/test_model.py (hunk 3 above)
Comment (ethanwharris) [reply tests/core/test_model.py:311]: > Or at least remove the initial attribute name within the task.

Definitely, I think there are a few enhancements here that would make it easier for users to know which layer to choose for embedding. It would also be cute if we could add a `backbone` layer or something that just goes through everything except the model head, but we may get that for free due to our naming conventions.

I'll leave these enhancements out of this PR as I need to play around with it a bit to figure out what the challenges are, but definitely some things we should look in to!

File: tests/core/test_model.py (hunk 2 above)
Comment (ethanwharris) [reply tests/core/test_model.py]: Good call! Done :smiley:
